src/os_pcap_nodelet.cpp

```diff
@@ -0,0 +1,252 @@
+                    lidar_packets->read([this](const uint8_t* buffer) {
+                        on_lidar_packet_msg(buffer);
+                    });
+                }
+
+                NODELET_DEBUG("lidar_packets_processing_thread done.");
+            });
+    }
+
+    void stop_packet_processing_threads() {
+        NODELET_DEBUG("stopping packet processing threads.");
+
+        if (imu_packets_processing_thread->joinable()) {
+            imu_packets_processing_thread_active = false;
+            imu_packets_processing_thread->join();
+        }
+
+        if (lidar_packets_processing_thread->joinable()) {
+            lidar_packets_processing_thread_active = false;
+            lidar_packets_processing_thread->join();
+        }
+    }
+
+    void on_lidar_packet_msg(const uint8_t* raw_lidar_packet) {
+        // copying the data from queue buffer into the message buffer
+        // this can be avoided by constructing an abstraction where
+        // OusterSensor has its own RingBuffer of PacketMsg but for
+        // now we are focusing on optimizing the code for OusterDriver
+        std::memcpy(lidar_packet.buf.data(), raw_lidar_packet,
+                    lidar_packet.buf.size());
+        lidar_packet_pub.publish(lidar_packet);
+    }
+
+    void on_imu_packet_msg(const uint8_t* raw_imu_packet) {
+        // copying the data from queue buffer into the message buffer
+        // this can be avoided by constructing an abstraction where
+        // OusterSensor has its own RingBuffer of PacketMsg but for
+        // now we are focusing on optimizing the code for OusterDriver
+        std::memcpy(imu_packet.buf.data(), raw_imu_packet,
+                    imu_packet.buf.size());
+        imu_packet_pub.publish(imu_packet);
+    }
+
+    void read_packets(PcapReader& pcap, const sensor::packet_format& pf) {
+        size_t payload_size = pcap.next_packet();
+        auto packet_info = pcap.current_info();
+        while (payload_size) {
+            auto start = std::chrono::high_resolution_clock::now();
+            if (packet_info.dst_port == info.config.udp_port_imu) {
+                imu_packets->write_overwrite(
+                    [this, &pcap, &pf, &packet_info](uint8_t* buffer) {
+                        std::memcpy(buffer, pcap.current_data(),
+                                    pf.imu_packet_size);
+                    });
+            } else if (packet_info.dst_port == info.config.udp_port_lidar) {
+                lidar_packets->write_overwrite(
+                    [this, &pcap, &pf, &packet_info](uint8_t* buffer) {
+                        std::memcpy(buffer, pcap.current_data(),
+                                    pf.lidar_packet_size);
+                    });
+            } else {
+                std::cout << "unknown packet" << std::endl;
+            }
+            auto prev_packet_ts = packet_info.timestamp;
+            payload_size = pcap.next_packet();
```

Comment:
You might be able to flip this loop to not duplicate these but :shrug:

Reply:
what part is duplicated? the `next_packet` call?

Reply:
Yeah, and the one below it. Theres a pair above the loop and right here.

Reply:
I don't think it would be possible to flip, since I need info about of previous and current timestamp

Reply:
No worries, just wanted to check.